Upgrade: Reduce number of false-positive migrations of the important modifier

philipp-spiess opened the pull request:
The important candidate migration is one of the most broad we have since it matches for any utility that are prefixed with an exclamation mark.

When running the codemodes on our example projects, we noticed that this was instead creating false-positives with candidates used in code positions, e.g: 


```ts
export default {
  shouldNotUse: !border.shouldUse,
}
```

To prevent false-positives, this PR adds a heuristics to detect wether or not a candidate is used in a non-code position. We do this by checking the character before and after the modifier and only allow quotes or spaces.

This can cause candidates to not migrate that are valid Tailwind CSS classes, e.g.: 

```ts
let classNames = `!underline${isHovered ? ' font-bold' : ''}`
```

This, however, is not a big issue since v4 can parse the v3 important prefix too.

philipp-spiess commented:
* **#14737** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14737?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @philipp-spiess and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

adamwathan commented on packages/@tailwindcss-upgrade/src/template/codemods/important.test.ts:
```diff
@@ -15,5 +15,25 @@ test.each([
+
+  expect(
+    important(designSystem, {}, '!border', {
+      contents: `let notBorder = !border\n`,
```
If it's super easy, could we make this work with trailing whitespace too? Like `!border    \n`, for people not using Prettier or who don't have their editor configured to trim trailing whitespace and have stuff like this in there by mistake.

If it's too hard no big deal though.

adamwathan commented on CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - _Upgrade (experimental)_: Minify arbitrary values when printing candidates ([#14720](https://github.com/tailwindlabs/tailwindcss/pull/14720))
 - _Upgrade (experimental)_: Ensure legacy theme values ending in `1` (like `theme(spacing.1)`) are correctly migrated to custom properties ([#14724](https://github.com/tailwindlabs/tailwindcss/pull/14724))
 - _Upgrade (experimental)_: Migrate arbitrary values to bare values for the `from-*`, `via-*`, and `to-*` utilities ([#14725](https://github.com/tailwindlabs/tailwindcss/pull/14725))
+- _Upgrade (experimental)_: Reduce the number of false-positive migration of the important modifier (e.g. `!border` to `border!`) ([#14737](https://github.com/tailwindlabs/tailwindcss/pull/14737))
```
```suggestion
- _Upgrade (experimental)_: Don't migrate important modifiers that are actually logical negations (e.g. `let foo = !border` to `let foo = border!`) ([#14737](https://github.com/tailwindlabs/tailwindcss/pull/14737))
```

adamwathan commented on packages/@tailwindcss-upgrade/src/template/codemods/important.ts:
```diff
@@ -19,9 +19,44 @@ export function important(
+      // The important migration is one of the most broad migrations with a high
+      // potential of matching false positives since `!` is a valid character in
+      // most programming languages. Since v4 is technically backward compatible
+      // to v3 in that it can read `!` in the front of the utility too, we err
```
```suggestion
      // with v3 in that it can read `!` in the front of the utility too, we err
```

thecrypticace commented on packages/@tailwindcss-upgrade/src/template/codemods/important.ts:
```diff
@@ -19,9 +19,44 @@ export function important(
+        ) {
+          continue
+        }
+      }
```
Could we in theory scan backwards and forwards to look for quotes _on the same line? I feel like this logic may be too restrictive.

philipp-spiess replied:
liked that, implemented 😎

thecrypticace replied:
I think that change disappeared — don't see it in the diff 😬

philipp-spiess commented on packages/@tailwindcss-upgrade/src/template/codemods/important.ts:
```diff
@@ -19,9 +19,50 @@ export function important(
+  },
 ): string {
-  for (let candidate of parseCandidate(rawCandidate, designSystem)) {
+  nextCandidate: for (let candidate of parseCandidate(rawCandidate, designSystem)) {
     if (candidate.important && candidate.raw[candidate.raw.length - 1] !== '!') {
+      // The important migration is one of the most broad migrations with a high
+      // potential of matching false positives since `!` is a valid character in
+      // most programming languages. Since v4 is technically backward compatible
+      // with v3 in that it can read `!` in the front of the utility too, we err
+      // on the side of caution and only migrate candidates that we are certain
+      // are inside of a string.
+      if (location) {
+        let isQuoteBeforeCandidate = false
+        for (let i = location.start - 1; i >= 0; i--) {
+          let char = location.contents.at(i)!
+          if (char === '\n') {
+            break
+          }
+          if (isQuote(char)) {
+            isQuoteBeforeCandidate = true
+            break
+          }
+        }
+
+        let isQuoteAfterCandidate = false
+        for (let i = location.end; i < location.contents.length; i++) {
+          let char = location.contents.at(i)!
+          if (char === '\n') {
+            break
+          }
+          if (isQuote(char)) {
+            isQuoteAfterCandidate = true
+            break
+          }
+        }
+
+        if (!isQuoteBeforeCandidate || !isQuoteAfterCandidate) {
+          continue nextCandidate
```
```suggestion
  for (let candidate of parseCandidate(rawCandidate, designSystem)) {
    if (candidate.important && candidate.raw[candidate.raw.length - 1] !== '!') {
      // The important migration is one of the most broad migrations with a high
      // potential of matching false positives since `!` is a valid character in
      // most programming languages. Since v4 is technically backward compatible
      // with v3 in that it can read `!` in the front of the utility too, we err
      // on the side of caution and only migrate candidates that we are certain
      // are inside of a string.
      if (location) {
        let isQuoteBeforeCandidate = false
        for (let i = location.start - 1; i >= 0; i--) {
          let char = location.contents.at(i)!
          if (char === '\n') {
            break
          }
          if (isQuote(char)) {
            isQuoteBeforeCandidate = true
            break
          }
        }

        let isQuoteAfterCandidate = false
        for (let i = location.end; i < location.contents.length; i++) {
          let char = location.contents.at(i)!
          if (char === '\n') {
            break
          }
          if (isQuote(char)) {
            isQuoteAfterCandidate = true
            break
          }
        }

        if (!isQuoteBeforeCandidate || !isQuoteAfterCandidate) {
          continue
```

philipp-spiess commented:
### Merge activity

* **Oct 22, 12:07 PM EDT**: The merge label 'merge-queue' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=tailwindlabs&repo=tailwindcss) once it meets the requirements.

philipp-spiess committed:
Upgrade: Reduce number of false-positive migrations of the important modifier (#14737)

The important candidate migration is one of the most broad we have since it matches for any utility that are prefixed with an exclamation mark.

When running the codemodes on our example projects, we noticed that this was instead creating false-positives with candidates used in code positions, e.g:

```ts
export default {
  shouldNotUse: !border.shouldUse,
}
```

To prevent false-positives, this PR adds a heuristics to detect wether or not a candidate is used in a non-code position. We do this by checking the character before and after the modifier and only allow quotes or spaces.

This can cause candidates to not migrate that are valid Tailwind CSS classes, e.g.:

```ts
let classNames = `!underline${isHovered ? ' font-bold' : ''}`
```

This, however, is not a big issue since v4 can parse the v3 important prefix too.
